MAINT: stats: ensure functions work on non-default device

mdhaber opened the pull request:
#### Reference issue
Toward gh-22680

#### What does this implement/fix?
Makes adjustments needed to ensure that `stats` functions work when arrays are not on the default device. Adds tests for device and dtype preservation.

#### Additional information
To do:
- [x] add `device` to `marray` `asarray`
- [ ] ~~preserve device in `special/_support_alternative_backends.py`?~~ *A lot of work has been done on `_support_alternative_backends` since opening this; I don't think I need to do anything here.*
- [ ] ~~ensure common device in `xp_promote`?~~ *Sidestepped. Our private `_asarray` needs to support `device`.*
- [x] add tests for remaining functions

mdhaber committed:
WIP: stats: ensure functions work on non-default device

[skip ci]

mdhaber commented on scipy/stats/tests/test_device_dtype.py:
```diff
@@ -0,0 +1,138 @@
+     (stats.skew, {}),
+     (stats.skew, {'bias': False}),
+     (stats.kurtosis, {}),
+     (stats.kurtosis, {'bias': False}),git
```
```suggestion
     (stats.kurtosis, {'bias': False}),
```

rgommers commented:
Thanks Matt. This looks pretty good to me already.

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into gh22680

mdhaber committed:
MAINT: stats: post-merge fixups

mdhaber committed:
TST: stats: add dtype/device tests

mdhaber committed:
STY: stats: fix lint issues

mdhaber committed:
TST: stats: address test failures

mdhaber committed:
TST: stats: add tests of quantile and directional_stats

mdhaber committed:
MAINT: stats: skip marray tests

mdhaber committed:
MAINT: stats: address test failures

mdhaber committed:
STY: stats: fix lint

mdhaber commented on scipy/stats/_morestats.py:
```diff
@@ -987,7 +989,8 @@ def boxcox_llf(lmb, data, *, axis=0, keepdims=False, nan_policy='propagate'):
 def _boxcox_llf(data, axis=0, *, lmb):
     xp = array_namespace(data)
-    lmb, data = xp_promote(lmb, data, force_floating=True, xp=xp)
+    dtype = xp_result_type(lmb, data, force_floating=True, xp=xp)
+    data = xp.asarray(data, dtype=dtype)
```
It would be nice if `xp_promote` could convert Python scalar `lmb` to an array on the right device, but `xp_promote` uses our `_asarray`, which doesn't support `device`. I'd like to leave it as-is in this PR, but it's something to keep in mind in gh-22049.

mdhaber commented on scipy/stats/tests/test_marray.py:
```diff
@@ -7,6 +7,7 @@
 from scipy.stats._axis_nan_policy import _axis_nan_policy_factory
 
 
+pytest.skip(allow_module_level=True)  # temporary until `marray` supports `device`
```
`marray` 0.0.9 added `device`, but broke mdhaber/marray#123, so I pinned us at 0.0.8. I'll fix that soon (thoughts on the best way to do that?) and remove these skips.

mdhaber committed:
MAINT: unpin marray to use latest release; run marray tests

rgommers commented on scipy/conftest.py:
```diff
@@ -440,6 +440,10 @@ def test_device(xp, devices):
+    if is_torch(xp):
+        devices = xp.__array_namespace_info__().devices()
+        # open an issue about this - cannot branch based on `any`/`all`?
+        return (device for device in devices if device.type != 'meta')
```
Related to https://github.com/data-apis/array-api/issues/945 it looks like. `'meta'` backend is brand new and WIP I think - do we need to worry about it here if CI is green?

mdhaber replied:
I'm guessing not many existing tests exercise it because only tests that use the `devices` fixture see it, right? In any case, it causes several of these new tests to fail. 

I don't think I can make all the functions work with it in this PR, so currently I used this like to skip it everywhere. But I'm open to thoughts on a more targeted way to skip or xfail the offending tests.

rgommers replied:
> I'm guessing not many existing tests exercise it because only tests that use the `devices` fixture see it, right?

Yes, that seems right. And I think the choice you made here for skipping at the scipy-global level is the right approach for now.

crusaderky replied:
Supporting the torch `meta` device requires 
1. adding it to `array_api_compat.is_lazy_array`
2. switching scipy to use array-api-extra `xp_assert_equal` et al. to benefit from https://github.com/data-apis/array-api-extra/pull/300.

rgommers commented on scipy/spatial/distance.py:
```diff
@@ -2294,7 +2294,8 @@ def pdist(X, metric='euclidean', *, out=None, **kwargs):
     if X.ndim != 2:
-        raise ValueError(f'A 2-dimensional array must be passed. (Shape was {X.shape}).')
+        raise ValueError('A 2-dimensional array must be passed. '
+                         f'(Shape was {X.shape}).')
```
It'd be nice to drop the changes to this file, since it's style-only in a different module. Can be done in a separate PR if this bothers your local setup somehow.

mdhaber replied:
Yeah it came up with `spin lint`. I'll take it out.

rgommers commented on scipy/stats/_morestats.py:
```diff
@@ -2950,7 +2954,7 @@ def bartlett(*samples, axis=0):
     T = numer / denom
 
-    chi2 = _SimpleChi2(xp.asarray(k-1))
+    chi2 = _SimpleChi2(xp.asarray(k-1, dtype=dtype))
```
Looks right. This PR looks about ready to merge.

mdhaber committed:
MAINT: stats: adjustments per review

mdhaber commented:
Removed changes to `distance.py` and added the one missing use of `device` to `bartlett`. Since `meta` is new and essentially untested, I can open an issue for it. If we want to support it, I think we need a help function to look for it (like `is_meta_array`, since it will require special cases like lazy arrays do), and I think the test skip/`xp_capabilities` system needs to be updated to deal with it.

rgommers committed:
Merge branch 'main' into gh22680

rgommers committed:
MAINT: stats: some more device/dtype fixes

rgommers committed:
STY: fix a linter complaint

rgommers commented:
> Since `meta` is new and essentially untested, I can open an issue for it. If we want to support it, I think we need a help function to look for it (like `is_meta_array`, since it will require special cases like lazy arrays do), and I think the test skip/`xp_capabilities` system needs to be updated to deal with it.

An issue would be great. My current impression is that this is low-priority, and we should leave it completely disabled until we have dealt with JAX/Dask more. Because the PyTorch `'meta'` device has similar requirements to them.

rgommers reviewed (approved):
Thanks @mdhaber. I think this looks good in its current form, let's squash-merge when CI comes back green I'd say.

lucascolley committed:
[skip ci]

rgommers commented on scipy/stats/tests/test_device_dtype.py:
```diff
@@ -0,0 +1,240 @@
+                                    'normal_unbiased', 'harrell-davis'])
+@pytest.mark.parametrize('dtype', dtypes)
+def test_quantile(method, dtype, xp, devices):
+    if (is_array_api_strict(xp) or is_torch(xp)) and method == 'harrell-davis':
+        pytest.skip("'harrell-davis' not currently not supported on GPU.")
```
Good point. It's easier to write it with `if ...` rather than `pytest.param` (no need to remember unusual `pytest` syntax), but the `pytest.param` method is a bit more concise and gives better messages.

I'm happy for this to go in either way though, before it accumulates conflicts again.

lucascolley replied:
actually, it looks like `cpu_only=True` does not skip array-api-strict's `device1` as that comes via the `devices` fixture, not the `xp` fixture.

@crusaderky is there a better solution already available here?

crusaderky replied:
What you're asking for is a flag for all backends that have devices other than CPU. Which doesn't exist, and it's a tad too niche IMHO.
Worth pointing out that, on a non-pixi deployment with CUDA installed,
```python
@pytest.mark.skip_xp_backends(cpu_only=True)
def test1(xp, devices):
    ...
```
will in fact run on torch GPU and JAX GPU.

lucascolley committed:
[skip ci]

crusaderky commented on scipy/stats/tests/test_device_dtype.py:
```diff
@@ -0,0 +1,240 @@
+@pytest.mark.parametrize('dtype', dtypes)
+def test_one_in_one_out(fun, kwargs, dtype, xp, devices):
+    if is_dask(xp) and fun == stats.variation:
+        pytest.skip("dtype inference failing in xpx.apply_where")
```
@mdhaber this is quite worrysome:
> E           TypeError("Multiple namespaces for array inputs: {<module 'scipy._lib.array_api_compat.numpy' [...], <module 'scipy._lib.array_api_compat.dask.array' [...]

mdhaber replied:
I'm going to look into it, but it wasn't worth holding up the whole PR. Dask can be very creative about how to fail with code that works for other backends.

rgommers replied:
> Dask can be very creative about how to fail with code that works for other backends.

Yeah I looked at this, and it's very much not obvious what is going on there. All the input variables seemed to be Dask arrays of the same dtype and device, and internally somewhere, one is converted to a numpy array.
